Add F# ( command dotnet-compile-fsc )

enricosada opened the pull request:
add command `dotnet-compile-fsc`

It's used with `"compilerName": "fsc"` in project.json

use`fsc.exe` from `Microsoft.FSharp.Compiler.netcore` package ( ref http://myget.org/gallery/fsharp-daily )

the FSharp.Core.netcore package is in the same feed

dnfclas commented:
Hi __@enricosada__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dnfclas commented:
__@enricosada__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

Alxandr commented:
How are you dealing with file order? You should have a test project with more than 1 source file.

agocke commented on src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json:
```diff
@@ -0,0 +1,31 @@
+    "dependencies": {
+        "NETStandard.Library": "1.0.0-rc2-23616",
+        
+        "Microsoft.Net.Compilers.netcore": "1.2.0-beta-20151215-01",
```
These are the Roslyn compilers -- you should remove it.

enricosada replied:
Ok

agocke commented on src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json:
```diff
@@ -0,0 +1,31 @@
+            "type": "build",
+            "version": "1.0.0-*"
+        },
+        "Microsoft.DotNet.Compiler.Common": "1.0.0-*"
```
I would expect a reference to a coreclr compatible F# compiler somewhere around here.

Alxandr replied:
At present, no such thing exists afaik.

enricosada replied:
No released version of coreclr fcs. 
We can use the desktop version installed to use dotnet cli to compile project.json. 
That can help existing projects right now to try dotnet cli

agocke replied:
Maybe something on http://myget.org/gallery/fsharp-daily would suffice.

Otherwise, I would consider this a blocker. F# needs to be delivered with dotnet-cli and it needs to run on CoreCLR.

enricosada replied:
added fsc coreclr from  http://myget.org/gallery/fsharp-daily

agocke commented on src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs:
```diff
@@ -0,0 +1,145 @@
+        private static Command RunFsc(string fscArgs)
+        {
+            // Locate CoreRun
+            return Command.Create("fsc", fscArgs);
```
This doesn't seem to locate CoreRun. :)

Also, dotnet-cli needs to bring its dependencies with it. You can't rely on `fsc` being installed on the machine.

enricosada replied:
I use fsc in PATH for now, no released version of fcs
At least i can use dotnet cli to compile existing f# project and try project.json
Is not the final coreclr ready implementation but is the starting point

enricosada replied:
Btw this comment is from original csc project

enricosada replied:
```csharp
// Locate CoreRun
```

this commet was from csc implementation ( i used that as a base )

agocke commented on src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs:
```diff
@@ -0,0 +1,145 @@
+            };
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                args.Add("--debug:full");
```
Does fsc support portable PDBs? If so, we should use that on other platforms.

enricosada replied:
Not portable pdb atm, in the todo list

agocke commented on src/Microsoft.DotNet.Tools.Compiler.Fsc/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Is this file necessary? I don't see anything interesting happening here.

enricosada replied:
Nope, i'll remove it

enricosada commented:
ok @agocke done! 

- fsc from netcore ( fsharp daily nuget feed )
- use file ordering ( with `compileFiles` in project.json ) in test

enricosada commented:
about build errors:

- osx, error on git clone

```
06:22:16  > git init /Users/dotnet-bot/j/workspace/dotnet_cli/debug_osx_prtest # timeout=10
06:22:16 
ERROR: Error cloning remote repo 'origin'
```

- win release:

```
06:36:14 Compiling Microsoft.DotNet.ProjectModel for DNXCore,Version=v5.0
06:36:18 
06:36:18 Unhandled Exception: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
06:36:18    at Microsoft.Cci.ISymUnmanagedWriter2.Close()
```

i get error like that sometimes on build ( on master branch too, not only in my pr ) and on different projects. dunno. if i rerun `build.cmd` more times, it build correctly

enricosada commented:
btw @agocke thx for review!

agocke commented:
No problem :). My plane's about to take off but I'll see if I can try it out and get this merged ASAP.

agocke commented:
It looks like the FSharpTestApp and FSharpTestLibrary are not compiled in the smoke test -- we should probably add those to `smoke-test.sh`.

It also seems to fail with the following error on OS X:

```
Compiling FSharpTestApp for DNXCore,Version=v5.0

Unhandled Exception: System.ComponentModel.Win32Exception: No such file or directory
   at System.Diagnostics.Process.ResolvePath(String filename)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at Microsoft.DotNet.Cli.Utils.Command.Execute()
   at Microsoft.DotNet.Tools.Compiler.Fsc.Program.Main(String[] args)

Compilation failed.
    0 Warning(s)
    0 Error(s)
```

Have you tested this on a non-Windows machine?

enricosada commented:
I'll check tomorrow on my OSX but i think fcs doesnt work xplat atm. 
I'll add library to smoke test and check build log

enricosada commented:
- rebased to latest `HEAD`
- added `-help` ( like `dotnet-compile-csc` )
- use `--simpleresolution`
- disable pdb on osx ( doesnt work )
- move test projects in `test/FSharpTestProjects`, same structure as `test/TestProjects`, so we can use same integration tests i think ( changing only test directory )
- add CompileFail and TestAppWithArgs test
- locate and use `corerun` to execute `fsc.exe` 
- HACK: there is no FSharp.Core nuget atm, so i use the FSharp.Core.dll bundled with fsc.exe /cc @KevinRansom
- exe doesnt work ( error error FS0208 ) /cc @KevinRansom

KevinRansom commented:
The latest f# nuget package is here:

https://www.myget.org/feed/fsharp-daily/package/nuget/Microsoft.FSharp.Compiler.netcore/1.0.0-alpha-151218



From: Enrico Sada [mailto:notifications@github.com]
Sent: Monday, December 21, 2015 9:03 AM
To: dotnet/cli <cli@noreply.github.com>
Cc: Kevin Ransom <Kevin.Ransom@microsoft.com>
Subject: Re: [cli] Add F# ( command dotnet-compile-fsc ) (#538)


  *   rebased to latest HEAD
  *   added -help ( like dotnet-compile-csc )
  *   use --simpleresolution
  *   disable pdb on osx ( doesnt work )
  *   move test projects in test/FSharpTestProjects, same structure as test/TestProjects, so we can use same integration tests i think ( changing only test directory )
  *   add CompileFail and TestAppWithArgs test
  *   locate and use corerun to execute fsc.exe
  *   HACK: there is no FSharp.Core nuget atm, so i use the FSharp.Core.dll bundled with fsc.exe /cc @KevinRansom<https://github.com/KevinRansom>
  *   exe doesnt work ( error error FS0208 ) /cc @KevinRansom<https://github.com/KevinRansom>

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/cli/pull/538#issuecomment-166360000>.

enricosada commented:
@agocke i need to check osx, setting up environment, i'll try tonight

enricosada commented:
@KevinRansom i need a FSharp.Core nuget package, for add the FSharp.Core to library project and console project. 
i cant directly use the compiler nuget, let's speak in standup tonight :smile:

enricosada commented:
ok

- **compile to exe works!!** ( TestApp, TestAppWithArgs )
- use FSharp.Core coreclr nuget ( thx @KevinRansom )
- upgrate to latest alpha of fsharp compiler
- workaround FS0208 ( ref Microsoft/visualfsharp#824 )
- use default.win32manifest ( ref Microsoft/visualfsharp#825 )

enricosada commented:
ouch, all failed because

```
06:28:09 More than one runtime.json file has declared imports for 'any'
```

i see `runtimes\any\native` in `Microsoft.FSharp.Compiler.netcore.1.0.0-alpha-151218.nupkg` , but was there before, a new dependency added it?
I dont know how to fix it, locally was working before ( rebuilding from stratch atm to check )

/cc @KevinRansom @agocke @davidfowl

KevinRansom commented:
Compiler now has a dependency on fsharp.core

From: Enrico Sada [mailto:notifications@github.com]
Sent: Tuesday, December 22, 2015 6:47 AM
To: dotnet/cli <cli@noreply.github.com>
Cc: Kevin Ransom <Kevin.Ransom@microsoft.com>
Subject: Re: [cli] Add F# ( command dotnet-compile-fsc ) (#538)


ouch, all failed because

06:28:09 More than one runtime.json file has declared imports for 'any'

i see runtimes\any\native in Microsoft.FSharp.Compiler.netcore.1.0.0-alpha-151218.nupkg , but was there before, a new dependency added it?
I dont know how to fix it, locally was working before ( rebuilding from stratch atm to check )

/cc @KevinRansom<https://github.com/KevinRansom> @agocke<https://github.com/agocke> @davidfowl<https://github.com/davidfowl>

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/cli/pull/538#issuecomment-166633512>.

enricosada commented:
revert upgrade because `Microsoft.FSharp.Core.netcore` v1.0.0-alpha-151221 give an error with `NETStandard.Platform` v1.0.0-rc2-23616

ref Microsoft/visualfsharp#832

piotrpMSFT commented:
@enricosada can you squash the commits? Then we'll go ahead and merge.

enricosada commented:
@piotrpMSFT  sure, trying to build locally atm.

enricosada commented:
@piotrpMSFT done, tested locally after squash

enricosada commented:
```
[1m[33m/mnt/resource/j/workspace/dotnet_cli/debug_centos7.1_prtest/tools/MultiProjectValidator/tools/MultiProjectValidator/Program.cs(30,29): warning CS0168: The variable 'e' is declared but never used[39m[22m
08:10:31 
08:10:31 Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
08:10:31    at ProjectSanity.AnalysisRules.DependencyMismatch.DependencyInfo.Create(ProjectContext context, ```

i dont think centos build failures are about this pr.
